### PR TITLE
Add index-of

### PR DESCRIPTION
**Pull request by darkleaf**
Fixes https://github.com/weavejester/medley/issues/54

### CODE REVIEW

File: src/medley/core.cljc [hunk 1]
```diff
@@ -658,3 +658,10 @@
   {:added "1.4.0"}
   [x]
   (instance? #?(:clj java.util.regex.Pattern :cljs js/RegExp) x))
+
+(defn index-of
+  "Returns the index of the first occurrence of x in a sequential collection, coll, or -1 if not found."
+  [^java.util.List xs x]
```
Comment (darkleaf) [inline src/medley/core.cljc:664]: I haven't used ClojureScript in a long time, but it seems that it works even with the advanced level of optimization. I tested this by replacing :simple with :advanced in project.clj.

**Review by weavejester — CHANGES_REQUESTED** [review]

File: src/medley/core.cljc [hunk 2]
```diff
@@ -658,3 +658,11 @@
   {:added "1.4.0"}
   [x]
   (instance? #?(:clj java.util.regex.Pattern :cljs js/RegExp) x))
+
+(defn index-of
+  "Returns the index of the first occurrence of x in a sequential collection, coll,
```
Comment (tomdl89) [inline src/medley/core.cljc:663]: Docstring says `coll` but param is called `xs` so need to make them agree, one way or the other.

Comment (weavejester) [reply src/medley/core.cljc:663]: Ah, yep - `coll` is preferred in this case, since it tends to be what's used in many `clojure.core` functions.

**Comment by weavejester** [comment]
Thanks! I just thought of two more things, but pretty sure these are the last:

1. Can you add `{:added "1.9.0"}` metadata to the `index-of` function?
2. Can you change at least one of the tests so the index returned is something other than 1? Just to give us a little more coverage against potential regressions.

**Comment by weavejester** [comment]
Thanks for the updates. Can you wrap the docstring at 80 characters, as per a previous comment? e.g.

```clojure
(defn index-of
  "Returns the index of the first occurrence of the item in the sequential
  collection coll, or -1 if not found."
  {:added "1.9.0"}
  [^java.util.List coll item]
  (if (nil? coll)
    -1
    (.indexOf coll item)))
```

For some reason it's wrapped at a much smaller width. Once that's done I'll merge it in and cut a release.

**Comment by weavejester** [comment]
Thanks!

**Comment by NoahTheDuke** [comment]
This doesn't match clojure.string/index-of's behavior (return `nil` if not found). Is it intentional to have a mismatch?

**Comment by weavejester** [comment]
I assume the idea was to have it match the .indexOf method. However, it's a good point that the Clojure `clojure.string/index-of` function works differently, as it doesn't have to use primitives and also has nil-punning.

I think you're right though, @NoahTheDuke. I think it makes more sense for `index-of` to return `nil` rather than `-1`. I'll make the necessary changes.

**Comment by NoahTheDuke** [comment]
Sorry for not noticing until it got merged. 😓

**Pull request merged**